kubectl-cloudflow/verify/blueprint.go

```diff
@@ -0,0 +1,1007 @@
+	return fmt.Sprintf("Outlet%s is not compatible with inlet %s.", b.outletPortPath.ToString(), b.inletPath.ToString())
+}
+
+func (b IncompatibleSchema) InletPath() VerifiedPortPath {
+	return b.inletPath
+}
+
+func (b InvalidInletName) ToMessage() string {
+	return fmt.Sprintf("Inlet `%s` in streamlet `%s` is invalid. Names must consist of lower case alphanumeric characters and may contain '-' except for at the start or end.",
+		b.name, b.className)
+}
+
+func (b InvalidOutletName) ToMessage() string {
+	return fmt.Sprintf("Outlet `%s` in streamlet `%s` is invalid. Names must consist of lower case alphanumeric characters and may contain '-' except for at the start or end.",
+		b.name, b.className)
+}
+
+func (b InvalidPortPath) ToMessage() string {
+	return fmt.Sprintf("'%s' is not a valid path to an outlet or an inlet.", b.path)
+}
+
+func (b InvalidVolumeMountName) ToMessage() string {
+	return fmt.Sprintf("Volume mount `%s` in streamlet `%s` is invalid. Names must consist of lower case alphanumeric characters and may contain '-' except for at the start or end.",
+		b.name, b.className)
+}
+
+func (b NonAbsoluteVolumeMountPath) ToMessage() string {
+	return fmt.Sprintf("`%s` contains a volume mount `%s` with a non-absolute path (`%s`).", b.className, b.name, b.path)
+}
+
+func (b PortPathNotFound) ToMessage() string {
+	var end = "."
+	if b.suggestions != nil {
+		//TODO: refactor this to a mkString function
+		var suggestionsFormatted string = ""
+		for _, suggestion := range b.suggestions {
+			if suggestionsFormatted == "" {
+				suggestionsFormatted = suggestion.ToString()
+			} else {
+				suggestionsFormatted = suggestionsFormatted + " or " + suggestion.ToString()
+			}
+		}
+		end = fmt.Sprintf(", please try %s.", suggestionsFormatted)
+	}
+	return fmt.Sprintf("'%s' does not point to a known streamlet inlet or outlet%s", b.path, end)
+}
+
+func (b UnconnectedInlets) ToMessage() string {
+	var listFormatted = ""
+	for _, in := range b.unconnectedInlets {
+		if listFormatted == "" {
+			listFormatted = fmt.Sprintf("%s,%s", in.streamletRef, in.inlet.Name)
+		} else {
+			listFormatted = listFormatted + "," + fmt.Sprintf("%s,%s", in.streamletRef, in.inlet.Name)
+		}
+	}
+	return fmt.Sprintf("Inlets (%s) are not connected.", listFormatted)
+}
+
+type Blueprint struct {
+	images               map[string]domain.ImageReference
+	streamlets           []StreamletRef
+	connections          []StreamletConnection
+	streamletDescriptorsPerImage map[string][]StreamletDescriptor
+	globalProblems       []BlueprintProblem
+}
+
+// check for consistency between the image id mentioned in streamlets and the
+// image ids present in images section of the blueprint
+func checkStreamletImageConsistency(blueprint Blueprint) []BlueprintProblem {
+	var problems []BlueprintProblem
+	// check if the image ids in streamlets are present in the images section
+	for _, streamlet := range blueprint.streamlets {
+		if streamlet.imageId != nil {
```

Comment:
why would it be nil? do we support streamlets without an imageId part? Or maybe there is another error for this later on?

Reply:
Or I might have forgotten that we still want to support the old blueprint format where the images part is not present..

Reply:
yes that was the intention. backwards compatibility.